### PR TITLE
Kadykov pint

### PR DESCRIPTION
**Pull request by kadykov**
This commit fixes the following issue of the wrong behavior of `is_cupy` property with pint xarrays:
```pythonimport numpy as np
import xarray as xr
import cupy as cp
import pint_xarray
import cupy_xarray
from pint_xarray import unit_registry as ureg

cp_da = xr.DataArray(
    cp.linspace(0, 1, 11),
    coords=dict(
        x=np.linspace(0, 10, 11),
    ),
)

pint_da = cp_da.pint.quantify("meter")

print(f"is_cupy response: {pint_da.cupy.is_cupy}")
print(f"pint_da type {type(pint_da.pint.magnitude)}")
```
Currently, the output is:
```sh
is_cupy response: False
pint_da type <class 'cupy.ndarray'>
```

### CODE REVIEW

**Comment by dcherian** [comment]
Thanks!

Can you add a test please? You'll need to have `pint` as an optional dependency. SO we'll need to copy some of this code from Xarray: https://github.com/pydata/xarray/blob/6c5840e1198707cdcf7dc459f27ea9510eb76388/xarray/tests/__init__.py#L83

The test should look like:
``` python
@requires_pint
def test_is_cupy_pint():
	pass
```

File: cupy_xarray/accessors.py [hunk 1]
```diff
@@ -59,6 +61,8 @@ def as_cupy(self):
                 name=self.da.name,
                 attrs=self.da.attrs,
             )
+        if isinstance(self.da.data, pint_array_type):
+            return self.da.pint.dequantify().cupy.as_cupy().pint.quantify()
```
Comment (dcherian) [inline cupy_xarray/accessors.py:65]: Is it possible to simply replace the magnitude with a cupy array directly?

Comment (kadykov) [reply cupy_xarray/accessors.py]: Thank you for your advice!

I have changed the conversion from pint array to cupy array to the following one:
`data=(self.da.data.units * cp.asarray(self.da.data.magnitude))`
Now I realized that this solution does not work properly with pint arrays that were created with the `UnitRegistry` with the flag `force_ndarray_like` set to `True`. In this case, it gives the following error:
```sh
TypeError: Unsupported type <class 'numpy.ndarray'>
```
 And it seems that this flag is obligatory for working with `pint-xarray` module.

But so far, I don't have any better solution.

Also, I have added very basic tests, but they fail with the following error, and I don't really understand how to fix it:
```sh
E           cupy.cuda.compiler.CompileException: nvrtc: error: failed to open nvrtc-builtins64_118.dll.
E             Make sure that nvrtc-builtins64_118.dll is installed correctly.
```

Comment (dcherian) [reply cupy_xarray/accessors.py]: that looks like some cuda error.

@keewis what's the best way of converting a pint array magnitude to a cupy array

File: cupy_xarray/accessors.py [hunk 2]
```diff
@@ -52,20 +54,14 @@ def as_cupy(self):
...
-        return DataArray(
-            data=cp.asarray(self.da.data),
-            coords=self.da.coords,
-            dims=self.da.dims,
-            name=self.da.name,
-            attrs=self.da.attrs,
-        )
+        if isinstance(self.da.data, pint_array_type):
+            return self._as_dataarray(
+                data=(self.da.data.units * cp.asarray(self.da.data.magnitude)),
```
Comment (dcherian) [inline cupy_xarray/accessors.py:62]:
```suggestion
                data=pint.Quantity(cp.asarray(data.magnitude), units=data.units),
```

Comment (kadykov) [reply cupy_xarray/accessors.py]: Thank you for your suggestion, now `as_cupy()` works with pint DataArrays created from `pint_xarray` accessor.

File: cupy_xarray/accessors.py [hunk 3]
```diff
@@ -52,20 +54,16 @@ def as_cupy(self):
...
-                attrs=self.da.attrs,
             )
-        return DataArray(
-            data=cp.asarray(self.da.data),
-            coords=self.da.coords,
-            dims=self.da.dims,
-            name=self.da.name,
-            attrs=self.da.attrs,
-        )
+        if isinstance(self.da.data, pint_array_type):
```
Comment (dcherian) [inline cupy_xarray/accessors.py:60]: I think we have to handle the `pint(dask(numpy))` case explicitly.

Comment (kadykov) [reply cupy_xarray/accessors.py]: I agree, so I added the nested arrays support and basic tests

File: cupy_xarray/accessors.py [hunk 4]
```diff
@@ -24,9 +24,15 @@ def __init__(self, da):
     @property
     def is_cupy(self):
         """bool: The underlying data is a cupy array."""
-        if isinstance(self.da.data, dask_array_type):
-            return isinstance(self.da.data._meta, cp.ndarray)
-        return isinstance(self.da.data, cp.ndarray)
+        return self._get_datatype(self.da.data)
+
+    @classmethod
```
Comment (dcherian) [inline cupy_xarray/accessors.py:29]: Lets avoid the classmethods and just use a simple helper function outside the class. 

I like the Google style guide: https://google.github.io/styleguide/pyguide.html

File: cupy_xarray/tests/test_accessors.py [hunk 5]
```diff
@@ -1,10 +1,18 @@
 import numpy as np
 import pytest
 import xarray as xr
-from xarray.core.pycompat import dask_array_type
+from xarray.core.pycompat import DuckArrayModule
+from xarray.tests import requires_pint
```
Comment (dcherian) [inline cupy_xarray/tests/test_accessors.py:5]: Let's copy this over to `tests/__init__.py` from xarray

File: cupy_xarray/accessors.py (hunk 4 above)
Comment (kadykov) [reply cupy_xarray/accessors.py:29]: Thank you for the useful link, it is a real treasure chest of good practices!

File: cupy_xarray/tests/test_accessors.py [hunk 6]
```diff
@@ -64,3 +94,35 @@ def test_data_array_accessor_dask(tutorial_da_air_dask):
 
     da = da.cupy.as_numpy()
     assert not da.cupy.is_cupy
+
+
+@requires_pint
+def test_data_array_accessor_pint(tutorial_da_air_pint):
```
Comment (negin513) [inline cupy_xarray/tests/test_accessors.py:100]: This test checks this case:
pint + xr.DataArray --> cupy --> numpy

File: cupy_xarray/tests/test_accessors.py [hunk 7]
```diff
@@ -64,3 +94,35 @@ def test_data_array_accessor_dask(tutorial_da_air_dask):
...
+    assert da.cupy.is_cupy
+    assert isinstance(da.data, pint_array_type)
+
+    da = da.cupy.as_numpy()
+    assert not da.cupy.is_cupy
+    assert isinstance(da.data, pint_array_type)
+
+
+@requires_pint
+def test_data_array_accessor_pint_dask(tutorial_da_air_pint_dask):
```
Comment (negin513) [inline cupy_xarray/tests/test_accessors.py:115]: This test checks the following:
```
 # pint + Dask.DataArray --> cupy --> numpy
 ```

**Review by negin513 — COMMENTED** [review]
This pull request addresses an issue where the `is_cupy` property returns an incorrect value (False) for pint xarrays that are backed by CuPy arrays. The proposed modifications effectively resolve this issue and the changes are clear. Great job. 👍 

**Recommendations**
I have two recommendations regarding the tests:

* To ensure all the edge cases are adequately resolved and to prevent potential future problems, please ensure that the test cases cover all relevant edge cases and scenarios, as this will help to guarantee the robustness of the implemented fix. I added more comments to go over this. 

* I suggest improving the naming of test functions in the accessors module to better reflect their purpose and the functionality being tested.  I also want to  recommend adding docstrings to each test function, explaining the specific case or scenario that each test covers. Currently the name of the tests (and tests) are very confusing and this caused me some confusions on which cases are covered. 


The followings are the lists of tests of their purpose:

| Test Function Name                   | Description                           |
|--------------------------------------|---------------------------------------|
| test_data_set_accessor               | xr.dataset --> cupy ---> np           |
| test_data_array_accessor             | xr.DataArray --> cupy --> np          |
| test_data_array_accessor_dask        | dask.DataArray --> cupy --> np        |
| test_data_array_accessor_pint        | pint + xr.DataArray --> cupy --> np   |
| test_data_array_accessor_pint_dask   | pint + Dask.DataArray --> cupy --> np   |

**Comment by kadykov** [comment]
Thank you for your comments and for the good review of the changes.
I re-implemented the tests with the clearer naming and wider coverage of input data.

**Comment by kadykov** [comment]
Also closes this issue #31

**Comment by negin513** [comment]
Hey @dcherian , Do you have any additional comments for this PR? I think we can merge this PR.

**Comment by dcherian** [comment]
Sorry for the massive delay here. @kadykov 

I cleaned up the combinatorial  explosion of variables ;) but I don't have a machine to actually run the tests on. Can either you or @negin513 confirm that the tests  work please?

**Comment by dcherian** [comment]
Actually I have access now. fixing things...

File: cupy_xarray/tests/test_accessors.py [hunk 8]
```diff
@@ -1,66 +1,63 @@
...
+    assert not as_pint.cupy.is_cupy
+    cpda = as_pint.cupy.as_cupy()
+    if isinstance(cpda, xr.DataArray):
+        assert isinstance(cpda.data, pint.Quantity)
+    assert cpda.cupy.is_cupy
+
+    as_dask = as_pint.chunk()
+    if isinstance(as_dask, xr.DataArray):
+        assert isinstance(as_dask.data, pint.Quantity)
+        assert isinstance(as_dask.data.magnitude._meta, np.ndarray)
```
Comment (dcherian) [inline cupy_xarray/tests/test_accessors.py:58]: @keewis is  this expectation right for `DataArray(pint(np.ndarray)).chunk()`

Comment (keewis) [reply cupy_xarray/tests/test_accessors.py:58]: be aware that `DataArray.chunk()` simply passes the wrapped duck array to `dask.array.from_array` if it was not already a dask collection, it would produce a `dask(pint(np.ndarray))`. This is not what we want, so there's `DataArray.pint.chunk` to work around this. Basically, it dequantifies, applies `.chunk`, then requantifies.
